controller/src/test/java/io/pravega/controller/server/ZKControllerServiceMainTest.java

```diff
@@ -74,37 +112,61 @@ public void testZKSessionExpiry() {
             Assert.fail("Failed waiting for controllerServiceMain to get ready");
         }
 
+        MockControllerServiceStarter controllerServiceStarter = (MockControllerServiceStarter) controllerServiceMain.awaitServiceStarting();
         try {
-            controllerServiceMain.awaitServiceStarting().awaitRunning();
+            controllerServiceStarter.awaitRunning();
         } catch (IllegalStateException e) {
             log.error("Failed waiting for controllerServiceStarter to get ready", e);
             Assert.fail("Failed waiting for controllerServiceStarter to get ready");
             return;
         }
 
+        assertEquals(1, clientQueue.size());
+        CuratorFramework curatorClient = (CuratorFramework) clientQueue.poll().getClient();
         // Simulate ZK session timeout
         try {
-            controllerServiceMain.forceClientSessionExpiry();
+            curatorClient.getZookeeperClient().getZooKeeper().getTestable().injectSessionExpiration();
```

Comment:
If I'm not wrong, a ZK session expiration may happen before, during and after the callback execution. I understand that this is testing only the former case, right? Could we do something to extend this test and cover more cases in which a ZK session expiration may happen?

Reply:
You are right. We only test “before” here.  “After” is no problem as the curator still calls the callback if it has received the response from the server.  
“During” is something that is hard to simulate without adding hooks inside curator. What we can do is call them concurrently from this test in separate threads so that maybe sometimes “during” is simulated too. 

I did that experiment many times and   realised that curator behaves correctly in all three cases as long as we don’t call close on curator. Basically, the callback is always called if curator is in running state.